### ISSUE TITLE
[TRACKING ISSUE - API] Add Projects Endpoint

### ISSUE THREAD

**Issue by BethanyG** [codebuddies/backend#79]
##  Add a **`Osprojects`** api endpoint.

Goal: Create an API endpoint that allows users to submit open source projects to invite collaborators to.

### This is a **Tracking Issue**, individual task issues linked:

**`working branch`:**  `issue-79-add-projects-endpoint`
[x]  Create `osprojects` app, `apps.py` #81 
[x]  Create `osprojects` model & associated DB migration files #80 
[IN PROGRESS - BLOCKED by error]  Create `osprojects` admin view and register app with Django admin interface #84
[ ]  Create JSON spec for `osprojects` api endpoint #77 
[ ]  Create sample data/mocked data fixture for `osprojects` api endpoint #76 
[ ]  Create `osprojects` views & queries #82 
[ ]  Create `osprojects` serializer #83
[ ]  Create `osprojects` URLs  (`api\v1\projects`, and any others needed) #85
[ ]  Ensure associated unittests and test data are created for all code changes
[ ]  Document expected GET, POST, PATCH behavior and any other endpoint-specific docs 

[ ] Complete Unit Tests for #80 & #81  -- #179  -- **Tests were left off for these PRs, and need to be completed ASAP.**

**Comment by BethanyG** [comment]
Proposing this as a workflow for adding endpoints, and breaking down the tasks into smaller pieces.  

We'd make a tracking issue like this one for an endpoint, and then list out the different tasks/issues/points for discussion in their own issues that can be individually assigned to different people if needed.  

Not sure if I got the separation quite correct here - there may be dependencies in Django I am not remembering.  For instance, you can't make any views without a model...so models have to happen first..but views and models don't have to be done by the same person.  Likewise for serializers, URL maps or the admin view.  

Thoughts?  I is a bit intimidating to have responsibility for an entire endpoint.  That being said, it is also hard to coordinate multiple people working on one.

**Comment by lpatmo** [comment]
I pushed up a branch (empty right now) that folks could check out to commit to: `issue-79-add-projects-endpoint`

Thanks for listing all this out and the details here, @BethanyG!

**Comment by billglover** [comment]
I'd like to see if we can get into the habit of merging in smaller changes and tracking all the issues required to ship a feature as a milestone. For example, we can create mock data, merge it to master and continue shipping without needing to hold back those changes. We tend to trip up with larger PRs.

I do think shipping and entire endpoint requires some coordination, and probably some proactive encouragement to progress, assign, support and work on all required issues. Progress towards a milestone that can be released feels like it might provide a sense of working towards something.

I've set up milestones for bot Resources and Projects so you can see what I'm think. I can remove these if not desired.

**Comment by lpatmo** [comment]
> I've set up milestones for bot Resources and Projects so you can see what I'm thinking.

Thank you! Using this GitHub feature makes a lot of sense. Seeing all the tasks laid out visually in an issue is helpful too, but in the past I've also gotten into the problem of forgetting which issue is the 'master' issue. 

> I'd like to see if we can get into the habit of merging in smaller changes
💯

**Comment by BethanyG** [comment]
Stealing an idea from PyLadies, which names all their "epics" as [TRACKING ISSUE].  Can remove if this is too visually noisy.

**Comment by BethanyG** [comment]
Closing this.  It has had no activity since March, and it feels like development is moving in a different direction ATM.

**Comment by lpatmo** [comment]
Re-opening this to use as sandbox issue :)

**Comment by BethanyG** [comment]
@lpatmo  -- is @bkbuilt still working on this, or is this now something you are working on?  Do the associated issues also need to be reopened?

**Comment by lpatmo** [comment]
I believe @bkbuilt is focused on `issue-55-post-resource right` on the frontend now (PS let me know if you need help or want to pair on it again, @bkbuilt, and I will schedule a time to focus on it!)

@watchtheblur may be interested in pairing on the "adding a new endpoint" workflow, and we'll make sure to re-open the associated issues as we do each step. Thanks!

**Comment by BethanyG** [comment]
@lpatmo @watchtheblur -- tests were left off of PRs #169 and #170.  Rather than revert those and create problems for myself and others, I've opened a new issue for the missing tests.  Let's make sure we get them completed as soon as possible.  Otherwise, we can't call this complete.  Thanks!

**Comment by lpatmo** [comment]
Update: Tests for #169 weren't needed, but I started writing out some tests for the model (from #170) in this issue: https://github.com/codebuddies/backend/issues/179

**Comment by BethanyG** [comment]
_Not liking how you call me on my BS, stale bot!_  Removing the stale -- we are still working on this!!

**Comment by BethanyG** [comment]
@lpatmo - Are we going to actually add a projects endpoint at this stage, or can we just close this?

**Comment by BethanyG** [comment]
Still open.  Just sleeping.
